【Hackathon 9th Sprint No.87】feat: implement MultiDtypeGenerator for low-precision sample generation

Dayuxiaoshui opened the pull request:
### PR Category

Feature Enhancement

### Description

This PR implements `MultiDtypeGenerator`, a new plugin for generating low-precision computation graph samples from existing float32 samples. The generator supports creating float16, bfloat16, and float8 variants of samples, which is crucial for evaluating tensor compiler performance on mixed-precision workloads.

#### Key Features

1. **Multi-precision Support**
   - Generate low-precision samples (float16, bfloat16, float8) from existing float32 samples
   - Support batch generation with configurable dtype list via `dtype_list` parameter

2. **Automatic Dtype Conversion**
   - Automatically convert input tensor dtypes from float32 to target low-precision types in `input_meta.py`
   - Automatically convert weight tensor dtypes from float32 to target low-precision types in `weight_meta.py`
   - Intelligently preserves float32 precision for special weights that require it

3. **Smart Weight Preservation**
   - Keeps batch_norm parameters (`running_mean`, `running_var`, `scale`, `bias`) as float32 to avoid execution errors and precision degradation
   - Preserves layer_norm parameters as float32 when necessary
   - Maintains `num_batches_tracked` buffers as float32

4. **Autocast Integration**
   - Automatically injects `torch.autocast` context manager into `model.py` forward method
   - Handles operators that don't natively support low precision types
   - Automatically detects device type (cuda/cpu) from model code

5. **Metadata Management**
   - Updates `graph_net.json` with dtype information and generation metadata
   - Adds `multi_dtype_generated` tag to generated samples
   - Tracks source sample information via `generated_from` field

6. **Validation & Filtering**
   - Includes `MultiDtypeFilter` class for filtering out graphs/subgraphs that cannot execute properly with low precision
   - Validates generated samples can be loaded and parsed correctly
   - Removes invalid samples automatically

7. **Command-line Tool**
   - Provides `multi_dtype_generate.py` CLI tool for batch generation
   - Supports flexible configuration via command-line arguments
   - Easy integration with existing GraphNet workflows

8. **Compatibility**
   - Generated samples fully compatible with `test_compiler` evaluation workflow
   - Support for Agent code generation processes
   - Maintains all required file structure and metadata

#### Implementation Details

- **Core Module**: `graph_net/torch/multi_dtype_generator.py`
  - `MultiDtypeGenerator` class: Main generator implementation
  - `MultiDtypeFilter` class: Graph filtering mechanism
  
- **CLI Tool**: `graph_net/torch/multi_dtype_generate.py`
  - Command-line interface for batch processing

#### Usage Example

```bash
python -m graph_net.torch.multi_dtype_generate \
    --source-sample-path samples/ultralytics/yolov9t \
    --output-base-path /path/to/output \
    --dtype-list float16 bfloat16
```

#### Testing

- Tested on CV sample (yolov9t)
- Tested on NLP sample (t5-small)
- Verified batch_norm parameters correctly preserved as float32
- Confirmed autocast integration works correctly
- Validated generated samples can be loaded and executed

#### Background

According to statistics, GraphNet currently contains 2379 torch subgraph samples, with 97.9% (2330) using float precision and only 2.06% (49) using float16/bfloat16. This generator will help expand the low-precision sample coverage, enabling better evaluation of tensor compilers' mixed-precision optimization capabilities.

Dayuxiaoshui committed:
feat: implement MultiDtypeGenerator for low-precision sample generation

- Add MultiDtypeGenerator class to generate float16/bfloat16/float8 samples from float32 samples
- Automatically convert input and weight tensor dtypes
- Keep batch_norm parameters (running_mean, running_var, scale, bias) as float32
- Add torch.autocast context manager to model.py forward method
- Update graph_net.json metadata with dtype information
- Add sample validation mechanism
- Add MultiDtypeFilter for filtering invalid graphs
- Add command-line tool for batch generation
- Support test_compiler evaluation and Agent code generation workflows

Dayuxiaoshui committed:
refactor: optimize MultiDtypeGenerator code style and reduce nesting

- Refactor _modify_weight_meta to use single-pass traversal
- Extract _should_keep_weight_float32 helper method
- Simplify _ensure_autocast_import to reduce duplicate file writes
- Clean up unused variables in _validate_sample
- Apply black code formatting

lixinqi commented on graph_net/torch/multi_dtype_generate.py:
这个文件去掉吧。然后在`graph_net/torch/multi_dtype_generator.py`这个文件里写一个入口类，供run_model调用

Dayuxiaoshui committed:
feat: implement MultiDtypeGenerator with FX Graph passes

- Add MultiDtypeGenerator for generating low-precision (float16/bfloat16/float8) samples
- Implement FX Graph passes for dtype conversion (no AST/string manipulation)
- Preserve BatchNorm/LayerNorm buffers as float32 for numerical stability
- Add autocast metadata for runtime mixed-precision support
- Thread-safe extraction with proper locking
- Comprehensive error handling and device detection
- Validated on CV (ResNet18) and NLP (BERT) samples

Architecture:
- graph_net/torch/multi_dtype_generator.py: Main generator with RunModelDecorator
- graph_net/torch/multi_dtype_passes/: FX Graph pass implementations
  - pass_base.py: Base class for dtype conversion passes
  - dtype_conversion_pass.py: Concrete dtype conversion implementation
  - autocast_wrapper_pass.py: Autocast metadata injection
  - pass_mgr.py: Pass manager

Features:
- Real FX Graph passes (not string replacement)
- Smart weight preservation (BN/LN buffers stay float32)
- Automatic device type detection
- Compatible with test_compiler and Agent workflows
- Extensible filter system for unsupported operations

Fixes:
- Correct val_map semantics in FX Graph rewriting
- Proper dtype detection (avoid false positives on int tensors)
- CPU fallback when CUDA unavailable
- Import organization and error handling

Dayuxiaoshui committed:
fix: resolve ruff linter errors in multi_dtype_passes

- Add explicit re-export in __init__.py for DtypeConversionPass
- Use TYPE_CHECKING for forward reference in pass_mgr.py
- Remove unused torch import in autocast_wrapper_pass.py
- Fix type annotation to use type[DtypeConversionPass]

lixinqi commented on graph_net/torch/dtype_gen_passes/naive_autocast_dtype_pass.py:
这里要写清楚naive_autocast_dtype_pass

lixinqi commented on graph_net/torch/multi_dtype_passes/dtype_conversion_pass.py:
```diff
@@ -0,0 +1,132 @@
+                    # Only rewrite if not in preserve list
+                    attr_name = node.target
+                    if not self.should_preserve_weight(str(attr_name)):
+                        return True
```
使用node_need_rewrite来抽象

lixinqi commented on graph_net/torch/multi_dtype_generator.py:
```diff
@@ -0,0 +1,361 @@
+        # Apply autocast wrapper pass
+        # This ensures operators that don't support low precision can still execute
+        device_type = self._detect_device_type(gm_copy)
+        autocast_pass = AutocastWrapperPass(target_dtype=dtype, device_type=device_type)
```
这里的代码太硬了。如果AutocastWrapperPass只在几个模型上有效，而在更多模型上无效。那么就面临一个麻烦，要么重构这里的代码，要么重构AutocastWrapperPass内部的代码。这么反复改，bug就是个常量。

你参考dim_gen_pass的处理方法，对于给定计算图G，你把对应的dtype的pass names写到 graph_net.json的graph_net.graph_net_json_file_util.kDataTypeGeneralizationPasses字段里。参考kDimensionGeneralizationPasses

lixinqi commented on graph_net/torch/multi_dtype_generator.py:
```diff
@@ -0,0 +1,361 @@
+}
+
+
+class RunModelDecorator:
```
这个类名暗含了你希望用graph_net.torch.run_model来启动。
但是最好用graph_net.model_path_handler来启动。

本类名应该改成InitDataTypeGeneralizationPasses。“Init”单词提示了会往graph_net.json里写东西

lixinqi commented on graph_net/torch/multi_dtype_generator.py:
```diff
@@ -0,0 +1,361 @@
+
+    def __call__(self, model):
+        """Apply multi-dtype generation using GraphExtractor."""
+        from graph_net.torch.extractor import extract
```
```python
def __call__(self, model_path):
    from graph_net.torch.fx_graph_cache_util import (
        parse_immutable_model_path_into_sole_graph_module,
    )
    traced_model : fx.GraphModule = parse_immutable_model_path_into_sole_graph_module(model_path)
```
以后你都可以这样避开那个烦人的extractor。

lixinqi commented on graph_net/torch/multi_dtype_generator.py:
```diff
@@ -0,0 +1,361 @@
+        }
+
+
+class GraphExtractor:
```
这个类的主体逻辑移动到上面那个类里。

Dayuxiaoshui committed:
feat: add data type generalization passes with FX Graph architecture

- Add InitDataTypeGeneralizationPasses class for testing and initializing dtype conversion passes
- Implement DtypeConversionPass base class and ConcretePass for dtype conversion
- Add AutocastWrapperPass for handling operators not supporting low precision
- Store applicable pass names in graph_net.json under kDataTypeGeneralizationPasses
- Use FX Graph passes instead of AST/string manipulation for robustness
- Add atomic file writing for graph_net.json to prevent corruption
- Add comprehensive weight preservation logic for BatchNorm/LayerNorm parameters
- Add RunModelPredicator integration for testing graph runnability
- Refine exception handling with specific exception types
- Auto-detect device type for autocast (cuda/cpu)
- Organize passes in multi_dtype_passes/ directory with proper management

This feature enables generating low-precision (float16/bfloat16/float8) samples
from float32 computation graphs, following the initializer pattern where passes
are tested offline and applied dynamically at runtime.

lixinqi commented on graph_net/torch/multi_dtype_passes/__init__.py:
当前目录改名为dtype_gen_passes

lixinqi commented on graph_net/torch/dtype_gen_passes/dtype_generalization_pass.py:
改名为dtype_generalization_pass

lixinqi replied:
conversion的意思是把旧的变成新的。但我们想说泛化的意思。

lixinqi commented on graph_net/torch/dtype_generalizer.py:
改名 dtype_generalizer.py

lixinqi reviewed (commented):
需要准备一个graph_net/test/dtype_gen_test.sh测试入口文件。确保别人能跑起来你的代码。

lixinqi commented:
至少应该有两个步骤：1）挑选可行的dtype_generalization_pass名称写到graph_net.json；2）按dtype_generalization_pass将当前model_path改写为新数据类型的样本。

Dayuxiaoshui committed:
Refactor dtype generalization to two-step architecture per reviewer feedback

lixinqi commented on graph_net/torch/dtype_generalizer.py:
```diff
@@ -0,0 +1,371 @@
+        temp_path = graph_net_json_path.with_suffix(".json.tmp")
+        with open(temp_path, "w") as f:
+            json.dump(metadata, f, indent=4)
+        temp_path.replace(graph_net_json_path)
```
这里统一用graph_net_json_file_util.update_json方法

lixinqi commented on graph_net/test/dtype_gen_test.sh:
```diff
@@ -0,0 +1,135 @@
+)
+CONFIG_INIT=$(echo "$config_json_str_init" | base64 -w 0)
+
+python3 -m graph_net.torch.run_model \
```
入口换成`python3 -m  graph_net.model_path_handler`

lixinqi replied:
model_path_handler支持批量处理，它是通过--model-path-list指定一个list文件来完成的。

lixinqi commented on graph_net/test/dtype_gen_test.sh:
```diff
@@ -0,0 +1,135 @@
+    "decorator_path": "$GRAPH_NET_ROOT/torch/dtype_generalizer.py",
+    "decorator_class_name": "InitDataTypeGeneralizationPasses",
+    "decorator_config": {
+        "dtype_list": ["float16", "bfloat16"]
```
应该加一个model_path_prefix参数，这个参数一般就是`$GRAPH_NET_ROOT`，配合后面model_path_handler 的--model-path-list参数从文本里批量处理model_path.

lixinqi commented on graph_net/test/dtype_gen_test.sh:
```diff
@@ -0,0 +1,135 @@
+    "decorator_path": "$GRAPH_NET_ROOT/torch/dtype_generalizer.py",
+    "decorator_class_name": "ApplyDataTypeGeneralizationPasses",
+    "decorator_config": {
+        "output_dir": "$OUTPUT_DIR"
```
加model_path_prefix

lixinqi commented on graph_net/test/dtype_gen_test.sh:
```diff
@@ -0,0 +1,135 @@
+CONFIG_APPLY=$(echo "$config_json_str_apply" | base64 -w 0)
+
+echo "[1/2] Generating CV samples..."
+python3 -m graph_net.torch.run_model \
```
换新入口文件`python3 -m graph_net.model_path_handler`

lixinqi commented on graph_net/test/dtype_gen_test.sh:
```diff
@@ -0,0 +1,135 @@
+    "decorator_class_name": "ApplyDataTypeGeneralizationPasses",
+    "decorator_config": {
+        "output_dir": "$OUTPUT_DIR"
+    }
```
参考sym_dim_generalizer，添加model_runnable_predicator_{path, class_name, config}。
生成的代码必须经过试跑

lixinqi commented on graph_net/test/dtype_gen_test.sh:
```diff
@@ -0,0 +1,135 @@
+echo "=========================================="
+echo "Verification"
+echo "=========================================="
+echo ""
```
这些代码没有必要了。因为上面有model_runnable_predicator

Dayuxiaoshui committed:
Fix integer tensor type detection in dtype generalization pass

Dayuxiaoshui committed:
Resolve merge conflicts: merge dtype and develop branches

Dayuxiaoshui committed:
Merge branch 'develop' into dtype

Dayuxiaoshui committed:
Fix code formatting: apply black formatter

Dayuxiaoshui committed:
Fix code formatting: add missing blank line

lixinqi commented on samples/timm/resnet18/graph_net.json:
```diff
@@ -4,5 +4,9 @@
     "num_nodes_required": 1,
     "source": "timm",
     "heuristic_tag": "computer_vision",
+    "data_type_generalization_passes": [
+        "dtype_generalization_pass_float16",
+        "dtype_generalization_pass_bfloat16"
+    ],
```
代码里没有这两个pass

lixinqi commented on graph_net/torch/dtype_generalizer.py:
```diff
@@ -0,0 +1,401 @@
+            {
+                "dtype": dtype,
+                "precision": dtype,
+                "generated_from_dtype_generalization": True,
```
这里定义得太随意了。
首先，graph_net.json顶层字段有什么样的健应该在graph_net_json_file_util里写成一个常量，作为事先声明。
其次，这个名字的语义太弱了，dtype, precision, generated_from_dtype_generalization这些字段都没有充分说明。

lixinqi commented on graph_net/graph_net_json_file_util.py:
```diff
@@ -1,16 +1,63 @@
 
 
-def update_json(model_path, field, value):
+def update_json(json_path: Union[str, Path], updates: dict) -> None:
```
这个api太随意了。参数updates在鼓励别人不加思考的塞字段过来。
和旧版的接口保持一致吧。
就算外界要调用多次，那也不麻烦。

Dayuxiaoshui committed:
Refactor: fix update_json API and add dtype generalization field constants

Dayuxiaoshui committed:
Fix: create separate pass files for each dtype to match graph_net.json

lixinqi commented on graph_net/test/dtype_gen_test.sh:
```diff
@@ -0,0 +1,104 @@
+echo "[1/2] Generating CV samples..."
+python3 -m graph_net.model_path_handler \
+    --model-path "timm/resnet18" \
+    --handler-config="$CONFIG_APPLY" || echo "Warning: CV generation failed"
```
这里的代码不够紧凑。你参考https://github.com/PaddlePaddle/GraphNet/blob/develop/graph_net/bash_templates/model_path_handler_sh.txt 改一下。
上同
下同

lixinqi commented on graph_net/torch/dtype_gen_passes/dtype_generalization_pass.py:
```diff
@@ -0,0 +1,155 @@
+                        args=(new_node, self.torch_dtype),
+                    )
+                    val_map[node] = converted_node
+                else:
```
把这里的代码包装到create_placeholder(node)函数里，create_placeholder定义在当前函数的本for循环之前。
此处代码简化为
```python
val_map[node] = create_placeholder(node)
```
下面的情况也类似

Dayuxiaoshui committed:
Refactor: simplify test script and extract helper functions in dtype pass
